bndl --endpoint argument

markusjura opened the pull request:
Adding argument `--endpoint` to `bndl` command. If specified, it replaces the existing endpoints of the given input, e.g. a bundle file

**Example**

```
bndl \
  test-bundle-v0-b1157bdcd7c9815100bdf2479704c66210c4dcd3d16f2519009ffafba935a4f9.zip \
  -o new-test-bundle-v0-b1157bdcd7c9815100bdf2479704c66210c4dcd3d16f2519009ffafba935a4f9.zip \
  --endpoint web \
  --component test-bundle bind-protocol http \
  --bind-port 5601 \
  --service-name web \
  --acl http:/ \
  --endpoint \
  --component test-bundle akka-remote
```

results into:

```
endpoints = {
  "web" = {
    bind-protocol = "http"
    bind-port     = 5601
    service-name  = "web"
    acls          = [
      {
        http = {
          requests = [
            {
              path-beg = "/"
            }
          ]
        }
      }
    ]
  },
  "akka-remote" = {
    bind-protocol = "tcp"
    bind-port     = 0
  }
}
```

Existing endpoints inside the specified component are removed.

An `--acl` option is converted into a request ACL with a `path-beg` attribute. The attributes `path` and `path-regex`, as well as `rewrite and `method` are not yet supported.

Specifying an endpoint as HOCON string is not yet supported but will be added in another commit.

All endpoint arguments have been grouped into a separate section in the `bndl --help`:

```
usage: bndl [-h] [-f {docker,oci-image,bundle}] [--image-tag IMAGE_TAG]
            [--image-name IMAGE_NAME] [-o [OUTPUT]] [--no-shazar]
            [--no-default-endpoints] [-e ENDPOINT [OPTS]]
            [--component COMPONENT] [--bind-protocol BIND_PROTOCOL]
            [--bind-port BIND_PORT] [--service-name SERVICE_NAME] [--acl ACL]
            [--with-check] [--component-description COMPONENT_DESCRIPTION]
            [--annotation ANNOTATIONS]
            [--compatibility-version [COMPATIBILITY_VERSION]]
            [--disk-space [DISK_SPACE]] [--memory [MEMORY]] [--name [NAME]]
            [--nr-of-cpus [NR_OF_CPUS]] [--roles [ROLES [ROLES ...]]]
            [--system [SYSTEM]] [--system-version [SYSTEM_VERSION]]
            [--tag TAGS] [--version [VERSION]]
            [source]

positional arguments:
...

endpoints:
  -e ENDPOINT [OPTS], --endpoint ENDPOINT [OPTS]
                        Endpoints to add to bundle.conf
                        If specified, existing endpoints are removed
                        Example: bndl --endpoint web --component web --bind-protocol http --service-name web --acl http:/subpath
  --component COMPONENT
                        Component to which an endpoint should be added
                        Used in conjunction with the --endpoint option
                        Required when specifying the --endpoint option
  --bind-protocol BIND_PROTOCOL
                        Bind protocol of an endpoint
                        Used in conjunction with the --endpoint option
                        When absent, tcp is used
  --bind-port BIND_PORT
                        Bind port of an endpoint
                        Used in conjunction with the --endpoint option
                        When absent, 0 is used, meaning a bind port is selected by ConductR
  --service-name SERVICE_NAME
                        Service name of an endpoint
                        Used in conjunction with the --endpoint option
                        When absent, the endpoint is not locatable via the service name
  --acl ACL             Request ACL of an endpoint
                        Used in conjunction with the --endpoint option
                        When absent, the endpoint will not be accessible via the Proxy

```

Note that the sub-argument `--component` is required so that we know at which component in the `bundle.conf` the endpoints need to be replaced. This could be simplified in the future, e.g. by replacing the endpoints of the first component if only one component exists in the `bundle.conf`. However, it is quite common that a `bundle.conf` contains at least two components, the service itself and the check component so I am not sure if we can find a good default here so that the user don't need to declare the `--component` argument for each `--endpoint` argument. 

@huntc @fsat @longshorej  Thoughts?

markusjura committed:
bndl —endpoint argument

Adding argument `—endpoint` to `bndl` command. If specified, it replaces the existing endpoints of the given input, e.g. a bundle file

**Example**

```
bndl  test-bundle-v0-b1157bdcd7c9815100bdf2479704c66210c4dcd3d16f2519009ffafba935a4f9.zip -o new-test-bundle-v0-b1157bdcd7c9815100bdf2479704c66210c4dcd3d16f2519009ffafba935a4f9.zip --endpoint web --component test-bundle bind-protocol http --bind-port 5601 --service-name web --acl http:/ --endpoint --component test-bundle akka-remote
```

results into:

```
endpoints = {
  "web" = {
    bind-protocol = "http"
    bind-port     = 5601
    service-name  = "web"
    acls          = [
      {
        http = {
          requests = [
            {
              path-beg = "/"
            }
          ]
        }
      }
    ]
  },
  "akka-remote" = {
    bind-protocol = "tcp"
    bind-port     = 0
  }
}
```

Existing endpoints inside the specified component are removed.

An `--acl` option is converted into a request ACL with a `path-beg` attribute. The attributes `path` and `path-regex`, as well as `rewrite and `method` are not yet supported.

Specifying an endpoint as HOCON string is not yet supported but will be added in another commit.

markusjura commented:
Verified that the `--endpoint` changes are correctly working when running the bundle:

```
$ bndl /Users/mj/.conductr/cache/bundle/test-bundle-v0-b1157bdcd7c9815100bdf2479704c66210c4dcd3d16f2519009ffafba935a4f9.zip --endpoint test --component test-bundle --service-name test --bind-protocol http --acl http:/test0 --acl http:/test1 | conduct load

$ conduct run test-bundle

$ curl http://192.168.10.1:9000/test0/something
The path was: 0/something

$ curl http://192.168.10.1:9000/test1/something
The path was: 1/something
```

markusjura commented:
**Checklist**

- [x] --endpoint parsing
- [x] Endpoint object with simple validation
- [x] HOCON parsing
- [x] Documentation
- [x] Unit tests

markusjura commented:
## Manual tests

**Load and run test bundle with modified endpoint**

```
$ bndl /Users/mj/.conductr/cache/bundle/test-bundle-v0-b1157bdcd7c9815100bdf2479704c66210c4dcd3d16f2519009ffafba935a4f9.zip --endpoint test --component test-bundle --service-name test --bind-protocol http --acl http:/test0 --acl http:/test1 | conduct load

$ conduct run test-bundle

$ curl http://192.168.10.1:9000/test0/something
The path was: 0/something

$ curl http://192.168.10.1:9000/test1/something
The path was: 1/something
```

**Specifying multiple endpoints**

```
bndl /Users/mj/.conductr/cache/bundle/test-bundle-v0-b1157bdcd7c9815100bdf2479704c66210c4dcd3d16f2519009ffafba935a4f9.zip -o /Users/mj/.conductr/cache/bundle/new/test-bundle-v0-b1157bdcd7c9815100bdf2479704c66210c4dcd3d16f2519009ffafba935a4f9.zip --endpoint web --service-name web --bind-protocol http --component test-bundle --acl http:/ --acl http:/subpath --endpoint akka-remote --component test-bundle
```

results into:

```
components {
  test-bundle {
    description = "test-bundle"
    file-system-type = "universal"
    start-command = [
      "test-bundle/bin/test-bundle"
      "-J-Xmx128m"
      "-J-Xms128m"
    ]
    endpoints {
      web {
        bind-protocol = "http"
        bind-port = 0
        service-name = "web"
        acls = [
          {
            http {
              requests = [
                {
                  path-beg = "/"
                }
                {
                  path-beg = "/subpath"
                }
              ]
            }
          }
        ]
      }
      akka-remote {
        bind-protocol = "tcp"
        bind-port = 0
      }
    }
  }
}
```

**Specifying tcp and udp endpoints**

```
bndl /Users/mj/.conductr/cache/bundle/test-bundle-v0-b1157bdcd7c9815100bdf2479704c66210c4dcd3d16f2519009ffafba935a4f9.zip -o /Users/mj/.conductr/cache/bundle/new/test-bundle-v0-b1157bdcd7c9815100bdf2479704c66210c4dcd3d16f2519009ffafba935a4f9.zip --endpoint tcp --component test-bundle --service-name tcp --acl tcp:[5000,5001] --acl "tcp:[5002, 5003]" --endpoint udp --component test-bundle --service-name udp --acl udp:[6000,6001] --acl "udp:[6002, 6003]"
```

results into:

```
components {
  test-bundle {
    description = "test-bundle"
    file-system-type = "universal"
    start-command = [
      "test-bundle/bin/test-bundle"
      "-J-Xmx128m"
      "-J-Xms128m"
    ]
    endpoints {
      tcp {
        bind-protocol = "tcp"
        bind-port = 0
        service-name = "tcp"
        acls = [
          {
            tcp {
              requests = [
                5000
                5001
                5002
                5003
              ]
            }
          }
        ]
      }
      udp {
        bind-protocol = "udp"
        bind-port = 0
        service-name = "udp"
        acls = [
          {
            udp {
              requests = [
                6000
                6001
                6002
                6003
              ]
            }
          }
        ]
      }
    }
  }
}
```

**Validation error: endpoint name not specified**

```
bndl /Users/mj/.conductr/cache/bundle/test-bundle-v0-b1157bdcd7c9815100bdf2479704c66210c4dcd3d16f2519009ffafba935a4f9.zip -o /Users/mj/.conductr/cache/bundle/new/test-bundle-v0-b1157bdcd7c9815100bdf2479704c66210c4dcd3d16f2519009ffafba935a4f9.zip --endpoint --component test-bundle
usage: bndl [-h] [-f {docker,oci-image,bundle}] [--image-tag IMAGE_TAG]
            [--image-name IMAGE_NAME] [-o [OUTPUT]] [--no-shazar]
            [--no-default-endpoints] [-e ENDPOINT [OPTS]]
            [--component COMPONENT] [--bind-protocol BIND_PROTOCOL]
            [--bind-port BIND_PORT] [--service-name SERVICE_NAME] [--acl ACL]
            [--with-check] [--component-description COMPONENT_DESCRIPTION]
            [--annotation ANNOTATIONS]
            [--compatibility-version [COMPATIBILITY_VERSION]]
            [--disk-space [DISK_SPACE]] [--memory [MEMORY]] [--name [NAME]]
            [--nr-of-cpus [NR_OF_CPUS]] [--roles [ROLES [ROLES ...]]]
            [--system [SYSTEM]] [--system-version [SYSTEM_VERSION]]
            [--tag TAGS] [--version [VERSION]]
            [source]
bndl: error: argument -e/--endpoint: expected one argument
```

**Validation error: component not specified**

```
bndl /Users/mj/.conductr/cache/bundle/test-bundle-v0-b1157bdcd7c9815100bdf2479704c66210c4dcd3d16f2519009ffafba935a4f9.zip -o /Users/mj/.conductr/cache/bundle/new/test-bundle-v0-b1157bdcd7c9815100bdf2479704c66210c4dcd3d16f2519009ffafba935a4f9.zip --endpoint web
bndl: argument --component is required when specifying argument --endpoint web
```


**Validation error: endpoint sub-argument specified without --endpoint argument**

```
bndl /Users/mj/.conductr/cache/bundle/test-bundle-v0-b1157bdcd7c9815100bdf2479704c66210c4dcd3d16f2519009ffafba935a4f9.zip -o /Users/mj/.conductr/cache/bundle/new/test-bundle-v0-b1157bdcd7c9815100bdf2479704c66210c4dcd3d16f2519009ffafba935a4f9.zip --component web
bndl: argument --component needs to be specified after the argument --endpoint
```


**Validation error: Ambigous bind protocol**

```
bndl /Users/mj/.conductr/cache/bundle/test-bundle-v0-b1157bdcd7c9815100bdf2479704c66210c4dcd3d16f2519009ffafba935a4f9.zip -o /Users/mj/.conductr/cache/bundle/new/test-bundle-v0-b1157bdcd7c9815100bdf2479704c66210c4dcd3d16f2519009ffafba935a4f9.zip --endpoint tcp --component test-bundle --service-name tcp --acl tcp:[5000,5001] --acl udp:[3000]
bndl: argument --bind-protocol is required when acls with different protocol families are specified
endpoint: tcp
acls: tcp:[5000,5001], udp:[3000]
```

fsat commented on conductr_cli/test/test_bndl_utils.py:
```diff
@@ -167,7 +168,23 @@ def test_load_bundle_args_into_conf(self):
+                    'name': 'akka-remote',
+                    'component': 'test-bundle',
+                    'bind-port': 2345
+                })
```
I'd suggest adding `Endpoint` which has TCP and UDP request ACLs so we have test coverage on this area.

markusjura replied:
Done

fsat commented on conductr_cli/bndl_main.py:
```diff
@@ -35,6 +38,40 @@ def run(argv=None):
+def set_endpoints(args):
+    def validate(endpoint):
+        if 'component' not in endpoint:
+            log.error('bndl: argument --component is required when specifying argument --endpoint')
```
Perhaps also display the name of the endpoint specified by `--endpoint` - this will help the user narrow down which endpoint argument is causing this error when there are multiple `--endpoint` specified.

fsat commented on conductr_cli/endpoint.py:
```diff
@@ -0,0 +1,115 @@
+        protocol_family = ProtocolFamily[protocol_family_string]
+
+        if other_parts.startswith('[') and other_parts.endswith(']'):
+            ports = [int(port.strip()) for port in other_parts[1:-1].split(',')]
```
Perhaps `ports = json.loads(other_parts)` can be used to simplify parsing?

fsat reviewed (changes requested):
@markusjura - great work! Overall code lgtm, I have some minor feedback on displaying error message when `--component` argument is not specified.

My request for change is mainly to do with adding unit test coverage for TCP and UDP acls.

I'd also request some manual test to be done for TCP and UDP acls as well, and recorded in the PR.

fsat commented on conductr_cli/endpoint.py:
```diff
@@ -0,0 +1,115 @@
+
+
+class RequestAcl:
+    request_mappings = []
```
`request_mappings` need to be declared inside `__init__` as:

```
self.request_mappings = []
```

I'm assuming you'd like to declare this as instance field, not static field.

Sorry for missing this earlier.

fsat commented on conductr_cli/endpoint.py:
```diff
@@ -0,0 +1,115 @@
+
+class TcpRequest:
+    protocol_family = ProtocolFamily.tcp
+    ports = []
```
This field should be deleted. I'm assuming you'd like an instance field, since it's declared inside `__init__`.

fsat commented on conductr_cli/endpoint.py:
```diff
@@ -0,0 +1,115 @@
+
+class UdpRequest:
+    protocol_family = ProtocolFamily.udp
+    ports = []
```
This field should be deleted. I'm assuming you'd like an instance field, since it's declared inside `__init__`.

fsat commented on conductr_cli/endpoint.py:
```diff
@@ -0,0 +1,115 @@
+    protocol_family = ProtocolFamily.http
+    method = None
+    path_beg = None
+    rewrite = None
```
These three fields should be moved into `__init__`:

```
self.method = None
self.path_beg = None
self.rewrite = None
```

markusjura committed:
Incorporate review feedback

markusjura committed:
Ambigous bind protocol

Changing the validation of the endpoint bind protocol.

If the argument `--bind-protocol` is set then we use it. Otherwise, we check if we can derive the bind protocol from the acls. If all acls contain only one protocol family, e.g. `http` then we set the bind protcol to this protocol family. If multiple protocol families in the acls exist then we raise an `AmbigousBindProtocolError` exception. If no acls have been specified then we set the bind protocol to `tcp` (the default).

On the `bndl` command, if the `AmbigousBindProtocolError` is thrown, we output the following validation error:

```
bndl: argument --bind-protocol is required when acls with different protocol families are specified
endpoint: tcp
acls: tcp:[5000,5001], udp:[3000]
```

huntc commented:
Nice work @markusjura !
